mlflow/entities/evaluation_dataset.py

```
@@ -0,0 +1,364 @@
+                    expectations[expectation.name] = expectation.value
+
+                trace_id = trace.info.trace_id
+
+                record_dict = {
+                    "inputs": inputs,
+                    "expectations": expectations,
+                    "source": {
+                        "source_type": DatasetRecordSourceType.TRACE.value,
+                        "source_data": {"trace_id": trace_id},
+                    },
+                }
+                record_dicts.append(record_dict)
+        else:
+            record_dicts = records
+
+        for record in record_dicts:
+            if not isinstance(record, dict):
+                raise MlflowException.invalid_parameter_value("Each record must be a dictionary")
+            if "inputs" not in record:
+                raise MlflowException.invalid_parameter_value(
+                    "Each record must have an 'inputs' field"
+                )
+
+        from mlflow.tracking._tracking_service.utils import _get_store
+
+        tracking_store = _get_store()
+
+        # TODO: Remove this hasattr check once all tracking stores implement
+        #  the evaluation dataset APIs
+        if hasattr(tracking_store, "upsert_evaluation_dataset_records"):
+            try:
+                tracking_store.get_evaluation_dataset(self.dataset_id)
+            except Exception as e:
+                raise MlflowException.invalid_parameter_value(
+                    f"Cannot add records to dataset {self.dataset_id}: Dataset not found in "
+                    f"current tracking store. The dataset must exist in the backend before "
+                    f"adding records."
+                ) from e
+
+            tracking_store.upsert_evaluation_dataset_records(
+                dataset_id=self.dataset_id, records=record_dicts, updated_by=self.last_updated_by
+            )
+            self._records = None
+        else:
+            if self._records is None:
```

Q: This is for temporary testing/dev right? We should raise an exception for non-supported tracking store I guess?